pom.xml

```diff
@@ -57,8 +57,6 @@
     <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
 
     <java.version>8</java.version>
-    <scala211.version>2.11.12</scala211.version>
```

Comment:
can we retain the default versions? although seems the default versions are never used. cc @pan3793  what's your opinion?

Reply:
we can retain `scala211.version` and `scala211.binary.version`, the default scala version will be used if neither `spark-X.Y` nor `flink-X.Y` profile were activated.

Reply:
> we can retain `scala211.version` and `scala211.binary.version`, the default scala version will be used if neither `spark-X.Y` nor `flink-X.Y` profile were activated.

+1. @RexXiong what do you think?

Reply:
I don't think so,  if you can't keep common/client/service...  modules same versions like Flink/Spark use in IDE, there would be dependency problem. So I suggest add a scala profile to change the versions when compile, or just retain the 2.11 default versions for guiding user compile.

Reply:
> 

I think what @pan3793 means is that, someone just opens the project to read code without selecting any profile, for example read modules that do not rely on a specific scala version, if we delete the default scala version, it will be a mess

Reply:
That's truly a problem, so the principle here is, we should make sure that the IDEA works well and that the maven commands are flexible enough.

How about

- define default `scala.version=2.12.15` and `scala.bianry.version=2.12` (as-is)
- remove `scala.version` and `scala.bianry.version` from engine-specific profiles, I mean `spark-X.Y` and `flink-X.Y`
- provide `scala-2.11` `scala-2.12` `scala-2.13` which defines `scala.version` and `scala.bianry.version`, so that users can use active them to overwrite the default properties, and that users are responsible to make sure the combination is valid.

Reply:
> That's truly a problem, so the principle here is, we should make sure that the IDEA works well and that the maven commands are flexible enough.
> 
> How about
> 
> * define default `scala.version=2.12.15` and `scala.bianry.version=2.12` (as-is)
> * remove `scala.version` and `scala.bianry.version` from engine-specific profiles, I mean `spark-X.Y` and `flink-X.Y`
> * provide `scala-2.11` `scala-2.12` `scala-2.13` which defines `scala.version` and `scala.bianry.version`, so that users can activate them to overwrite the default properties, and that users are responsible to make sure the combination is valid.

Agree, but we also can keep scala.version  and scala.bianry.version for engine-specific profiles as their default.

Reply:
> 

Alright, let's do it.

- retain global default versions
- retian default scala versions in engine-specific profiles
- provide scala profiles

Reply:
> > That's truly a problem, so the principle here is, we should make sure that the IDEA works well and that the maven commands are flexible enough.
> > How about
> > 
> > * define default `scala.version=2.12.15` and `scala.bianry.version=2.12` (as-is)
> > * remove `scala.version` and `scala.bianry.version` from engine-specific profiles, I mean `spark-X.Y` and `flink-X.Y`
> > * provide `scala-2.11` `scala-2.12` `scala-2.13` which defines `scala.version` and `scala.bianry.version`, so that users can activate them to overwrite the default properties, and that users are responsible to make sure the combination is valid.
> 
> Agree, but we also can keep scala.version and scala.bianry.version for engine-specific profiles as their default.

If we also retain default scala versions in engine-specific profiles, what will happen if someone selects both engine profile and scala profile?

Reply:
> If we also retain default scala versions in engine-specific profiles, what will happen if someone selects both engine profile and scala profile?

That's my concern, suppose it works, but it's still confusing.

Reply:
> 

It would be overwrited by version defines in scala profile

Reply:
> If we also retain default scala versions in engine-specific profiles, what will happen if someone selects both engine profile and scala profile?

Have a quick test, it depends on the declaration order.
https://gist.github.com/pan3793/9a076b93bad785a7489631a74fbc1dbb

Reply:
> > If we also retain default scala versions in engine-specific profiles, what will happen if someone selects both engine profile and scala profile?
> 
> Have a quick test, it depends on the declaration order. https://gist.github.com/pan3793/9a076b93bad785a7489631a74fbc1dbb

What if it is -Pscala-2.11,spark-3.3?

Reply:
> What if it is -Pscala-2.11,spark-3.3?

The order in command does not matter, what matter is the profile declaration in `pom.xml`.

But as I said above, it works but is still confusing. I suggest removing scala properties from engine profiles.

Reply:
I suddenly realized that these two lines are not global scala version, but properties used for flink plugin only, so this pr is fine. It's my fault, sorry...